pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,317 @@
+limitations under the License.
+*/
+
+package apiexportendpointslice
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/core"
+	corev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/cluster"
+	apisv1alpha1client "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/typed/apis/v1alpha1"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	corev1alpha1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/core/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	"github.com/kcp-dev/kcp/pkg/reconciler/committer"
+)
+
+const (
+	ControllerName = "kcp-apiexportendpointslice"
+)
+
+// NewController returns a new controller for APIExportEndpointSlices.
+// Shards and APIExports are read from the cache server
+func NewController(
+	apiExportEndpointSliceClusterInformer apisinformers.APIExportEndpointSliceClusterInformer,
+	shardClusterInformer corev1alpha1informers.ShardClusterInformer,
+	apiExportClusterInformer apisinformers.APIExportClusterInformer,
+	kcpClusterClient kcpclientset.ClusterInterface,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue: queue,
+		listAPIExportEndpointSlices: func() ([]*apisv1alpha1.APIExportEndpointSlice, error) {
+			return apiExportEndpointSliceClusterInformer.Lister().List(labels.Everything())
+		},
+		listShards: func() ([]*corev1alpha1.Shard, error) {
+			return shardClusterInformer.Lister().List(labels.Everything())
+		},
+		getAPIExportEndpointSlice: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExportEndpointSlice, error) {
+			return apiExportEndpointSliceClusterInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		getAPIExport: func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error) {
+			return indexers.ByPathAndName[*apisv1alpha1.APIExport](apisv1alpha1.Resource("apiexports"), apiExportClusterInformer.Informer().GetIndexer(), path, name)
+		},
+		getAPIExportKeys: func(index, keyString string) ([]string, error) {
+			return apiExportEndpointSliceClusterInformer.Informer().GetIndexer().IndexKeys(index, keyString)
+		},
+		getSliceByAPIExportKey: func(key string) (item *apisv1alpha1.APIExportEndpointSlice, exists bool, err error) {
```

Comment:
these are strange abstractions, or actually no abstractions. Instead do:
```
getSlicesForAPIExports(export *APIExport)
```
with two lookups in the `indexAPIExportEndpointSliceByAPIExport` index.

Reply:
replaced by using the informer directly in the enqueue logic and having a second struct for the reconcile as per your recommendation